### ISSUE TITLE
Size of filter doesn't match size of column

### ISSUE THREAD

**Issue by Tasselmi** [ClickHouse/ClickHouse#7714]
I am sure they are of equal length.

```
SELECT sum(length(filt) != length(s2))
FROM test

┌─sum(notEquals(length(filt), length(s2)))─┐
│                                        0 │
└──────────────────────────────────────────┘
```

BUt  `Size of filter doesn't match`
```
SELECT arrayFilter((x, y) -> y, s2, filt)
FROM test

→ Progress: 57.34 thousand rows, 101.03 MB (536.38 thousand rows/s., 945.04 MB/s.)  4%Received exception from server (version 19.16.2):
Code: 9. DB::Exception: Received from localhost:9000. DB::Exception: Size of filter doesn't match size of column..

0 rows in set. Elapsed: 0.420 sec. Processed 57.34 thousand rows, 101.03 MB (136.44 thousand rows/s., 240.39 MB/s.)
```

**Comment by Tasselmi** [comment]
the same syntax, yesterday it worked well, today it is broken.

```
↗ Progress: 55.93 thousand rows, 88.39 MB (462.26 thousand rows/s., 730.55 MB/s.)  4%Received exception from server (version 19.16.3):
Code: 9. DB::Exception: Received from localhost:9000. DB::Exception: Size of filter doesn't match size of column..

0 rows in set. Elapsed: 0.242 sec. Processed 55.93 thousand rows, 88.39 MB (231.11 thousand rows/s., 365.24 MB/s.)

```

**Comment by Tasselmi** [comment]
this works fine.
```
SELECT *
FROM test
LIMIT 5
```

but........
```
SELECT count()
FROM test

→ Progress: 55.93 thousand rows, 84.97 MB (479.51 thousand rows/s., 728.52 MB/s.)  4%Received exception from server (version 19.16.3):
Code: 9. DB::Exception: Received from localhost:9000. DB::Exception: Size of filter doesn't match size of column..

0 rows in set. Elapsed: 1.476 sec. Processed 55.93 thousand rows, 84.97 MB (37.89 thousand rows/s., 57.57 MB/s.)
```

what I am doing?
where am I ?
what's the meaning of life?
we are alone in the universe ?

**Comment by Tasselmi** [comment]
```
Stack trace:

0. 0x559cc1cd9f90 StackTrace::StackTrace() ?
1. 0x559cc1cd9d65 DB::Exception::Exception(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int) ?
2. 0x559cc5231e02 void DB::filterArraysImpl<unsigned char>(DB::PODArray<unsigned char, 4096ul, Allocator<false, false>, 15ul, 16ul> const&, DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul> const&, DB::PODArray<unsigned char, 4096ul, Allocator<false, false>, 15ul, 16ul>&, DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul>&, DB::PODArray<unsigned char, 4096ul, Allocator<false, false>, 15ul, 16ul> const&, long) ?
3. 0x559cc51d539b DB::ColumnString::filter(DB::PODArray<unsigned char, 4096ul, Allocator<false, false>, 15ul, 16ul> const&, long) const ?
4. 0x559cc3e1c5a3 DB::ArrayFilterImpl::execute(DB::ColumnArray const&, COW<DB::IColumn>::immutable_ptr<DB::IColumn>) ?
5. 0x559cc3e1f1c8 DB::FunctionArrayMapped<DB::ArrayFilterImpl, DB::NameArrayFilter>::executeImpl(DB::Block&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, unsigned long) ?
6. 0x559cc4e60984 DB::PreparedFunctionImpl::execute(DB::Block&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, unsigned long, bool) ?
7. 0x559cc578a9c1 DB::ExpressionAction::execute(DB::Block&, bool) const ?
8. 0x559cc578c0d5 DB::ExpressionActions::execute(DB::Block&, bool) const ?
9. 0x559cc561dc8a DB::ExpressionBlockInputStream::readImpl() ?
10. 0x559cc4eec457 DB::IBlockInputStream::read() ?
11. 0x559cc561dc6b DB::ExpressionBlockInputStream::readImpl() ?
12. 0x559cc4eec457 DB::IBlockInputStream::read() ?
13. 0x559cc561dc6b DB::ExpressionBlockInputStream::readImpl() ?
14. 0x559cc4eec457 DB::IBlockInputStream::read() ?
15. 0x559cc505b37f DB::ConcatBlockInputStream::readImpl() ?
16. 0x559cc4eec457 DB::IBlockInputStream::read() ?
17. 0x559cc56b80ad DB::Aggregator::execute(std::shared_ptr<DB::IBlockInputStream> const&, DB::AggregatedDataVariants&) ?
18. 0x559cc55e4f1c DB::AggregatingBlockInputStream::readImpl() ?
19. 0x559cc4eec457 DB::IBlockInputStream::read() ?
20. 0x559cc561dc6b DB::ExpressionBlockInputStream::readImpl() ?
21. 0x559cc4eec457 DB::IBlockInputStream::read() ?
22. 0x559cc561dc6b DB::ExpressionBlockInputStream::readImpl() ?
23. 0x559cc4eec457 DB::IBlockInputStream::read() ?
24. 0x559cc4ef137a DB::LimitBlockInputStream::readImpl() ?
25. 0x559cc4eec457 DB::IBlockInputStream::read() ?
26. 0x559cc4ee4f9b DB::AsynchronousBlockInputStream::calculate() ?
27. 0x559cc4ee5360 ? ?
28. 0x559cc1d26a6e ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::_List_iterator<ThreadFromGlobalPool>) ?
29. 0x559cc1d2707e ThreadFromGlobalPool::ThreadFromGlobalPool<ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::function<void ()>, int, std::optional<unsigned long>)::{lambda()#3}>(ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::function<void ()>, int, std::optional<unsigned long>)::{lambda()#3}&&)::{lambda()#1}::operator()() const ?
30. 0x559cc1d2453c ThreadPoolImpl<std::thread>::worker(std::_List_iterator<std::thread>) ?
31. 0x559cc7a401e0 ? ?
```

**Comment by filimonov** [comment]
Maybe some part get damaged for some reason? 

Can data be read with simple selects, like: `SELECT count(ignore(*)) from test`

Can you try to create a reproducible sample?

**Comment by d-ganchar** [comment]
I had the same problem when some mutations were stuck and any new mutation didn't work through. So I **KILL** them but the problem remained. Query example:
```
SELECT field1,
       field2,        
       count() counter
  FROM my_table
 WHERE toDate(dt) = toDate('2020-03-05') 
   AND field3 IN ('value1')
   AND field4 IN ('value2') 
 GROUP BY field1, field2

# DB::Exception: Size of filter doesn't match size of column. (version 20.1.2.4 (official build))
```

**Interesting thing:** the query works fine when I remove `AND field4 IN ('value2')` condition.
The problem disappeared when I copy data into a new table.
So in my case(I think) this happened due to many uncompleted mutations.

**Comment by d-ganchar** [comment]
Repeated again.
There is a table with data(`engine = MergeTree() PARTITION BY toMonth(some_col)`). Data is collected every day. I need to add a column. I did:

`ALTER TABLE my_tbl ADD COLUMN col_name Nullable(Float64) DEFAULT NULL;`

Past queries works fine. But I got the error when I try to add a new column into query(for example `... ORDER BY sum(col_name)`). The same query works fine when I move data into a new table.

**Comment by filimonov** [comment]
it looks it relates to prewhere and columns which has not value written in column yes (i.e. freshly added, absent in parts). Trying to reproduce...

**Comment by SaltTan** [comment]
https://github.com/ClickHouse/ClickHouse/issues/9132 was fixed in 20.1.7

**Comment by filimonov** [comment]
Fixed in https://github.com/ClickHouse/ClickHouse/pull/9612
